### ISSUE TITLE
Incorrect layout in SPIR-V passthrough

### ISSUE THREAD

**Issue by HeroicKatora** [gfx-rs/naga#1415]
The following shader produced a validation error after having passed through naga in wgpu.

```
    	Validation Error: [ UNASSIGNED-CoreValidation-Shader-InconsistentSpirv ] Object 0: handle = 0x7f7e50a46978, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x6bbb14 | SPIR-V module not valid: Structure id 18 decorated as Block for variable in Uniform storage class must follow relaxed uniform buffer layout rules: member 1 at offset 32 overlaps previous member ending at offset 63
      %PaintCoordinates = OpTypeStruct %mat4v2float %mat4v2float
```

The original layout was two members with size `32`, so the offset was correct.

[box.vert.v.gz](https://github.com/gfx-rs/naga/files/7205519/box.vert.v.gz)

<details>
<summary>spirv-dis</summary>

```glsl
; SPIR-V
; Version: 1.0
; Generator: Google Shaderc over Glslang; 10
; Bound: 113
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %vertPosition %_ %uv
               OpSource GLSL 450
               OpSourceExtension "GL_EXT_scalar_block_layout"
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %main "main"
               OpName %rect_selection_i1_ "rect_selection(i1;"
               OpName %i "i"
               OpName %rect_position_i1_ "rect_position(i1;"
               OpName %i_0 "i"
               OpName %PaintCoordinates "PaintCoordinates"
               OpMemberName %PaintCoordinates 0 "rect_selection"
               OpMemberName %PaintCoordinates 1 "rect_position"
               OpName %paint_coordinates "paint_coordinates"
               OpName %innerUv "innerUv"
               OpName %param "param"
               OpName %param_0 "param"
               OpName %vertPosition "vertPosition"
               OpName %param_1 "param"
               OpName %param_2 "param"
               OpName %innerPos "innerPos"
               OpName %param_3 "param"
               OpName %param_4 "param"
               OpName %param_5 "param"
               OpName %param_6 "param"
               OpName %glslPos "glslPos"
               OpName %gl_PerVertex "gl_PerVertex"
               OpMemberName %gl_PerVertex 0 "gl_Position"
               OpMemberName %gl_PerVertex 1 "gl_PointSize"
               OpMemberName %gl_PerVertex 2 "gl_ClipDistance"
               OpMemberName %gl_PerVertex 3 "gl_CullDistance"
               OpName %_ ""
               OpName %uv "uv"
               OpMemberDecorate %PaintCoordinates 0 ColMajor
               OpMemberDecorate %PaintCoordinates 0 Offset 0
               OpMemberDecorate %PaintCoordinates 0 MatrixStride 8
               OpMemberDecorate %PaintCoordinates 1 ColMajor
               OpMemberDecorate %PaintCoordinates 1 Offset 32
               OpMemberDecorate %PaintCoordinates 1 MatrixStride 8
               OpDecorate %PaintCoordinates Block
               OpDecorate %paint_coordinates DescriptorSet 0
               OpDecorate %paint_coordinates Binding 0
               OpDecorate %vertPosition Location 0
               OpMemberDecorate %gl_PerVertex 0 BuiltIn Position
               OpMemberDecorate %gl_PerVertex 1 BuiltIn PointSize
               OpMemberDecorate %gl_PerVertex 2 BuiltIn ClipDistance
               OpMemberDecorate %gl_PerVertex 3 BuiltIn CullDistance
               OpDecorate %gl_PerVertex Block
               OpDecorate %uv Location 0
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
%_ptr_Function_int = OpTypePointer Function %int
      %float = OpTypeFloat 32
    %v2float = OpTypeVector %float 2
         %10 = OpTypeFunction %v2float %_ptr_Function_int
%mat4v2float = OpTypeMatrix %v2float 4
%PaintCoordinates = OpTypeStruct %mat4v2float %mat4v2float
%_ptr_Uniform_PaintCoordinates = OpTypePointer Uniform %PaintCoordinates
%paint_coordinates = OpVariable %_ptr_Uniform_PaintCoordinates Uniform
      %int_0 = OpConstant %int 0
%_ptr_Uniform_v2float = OpTypePointer Uniform %v2float
      %int_1 = OpConstant %int 1
%_ptr_Function_v2float = OpTypePointer Function %v2float
%_ptr_Input_v2float = OpTypePointer Input %v2float
%vertPosition = OpVariable %_ptr_Input_v2float Input
       %uint = OpTypeInt 32 0
     %uint_0 = OpConstant %uint 0
%_ptr_Input_float = OpTypePointer Input %float
      %int_3 = OpConstant %int 3
      %int_2 = OpConstant %int 2
     %uint_1 = OpConstant %uint 1
    %float_2 = OpConstant %float 2
%_ptr_Function_float = OpTypePointer Function %float
    %float_1 = OpConstant %float 1
    %v4float = OpTypeVector %float 4
%_arr_float_uint_1 = OpTypeArray %float %uint_1
%gl_PerVertex = OpTypeStruct %v4float %float %_arr_float_uint_1 %_arr_float_uint_1
%_ptr_Output_gl_PerVertex = OpTypePointer Output %gl_PerVertex
          %_ = OpVariable %_ptr_Output_gl_PerVertex Output
    %float_0 = OpConstant %float 0
%_ptr_Output_v4float = OpTypePointer Output %v4float
%_ptr_Output_v2float = OpTypePointer Output %v2float
         %uv = OpVariable %_ptr_Output_v2float Output
       %main = OpFunction %void None %3
          %5 = OpLabel
    %innerUv = OpVariable %_ptr_Function_v2float Function
      %param = OpVariable %_ptr_Function_int Function
    %param_0 = OpVariable %_ptr_Function_int Function
    %param_1 = OpVariable %_ptr_Function_int Function
    %param_2 = OpVariable %_ptr_Function_int Function
   %innerPos = OpVariable %_ptr_Function_v2float Function
    %param_3 = OpVariable %_ptr_Function_int Function
    %param_4 = OpVariable %_ptr_Function_int Function
    %param_5 = OpVariable %_ptr_Function_int Function
    %param_6 = OpVariable %_ptr_Function_int Function
    %glslPos = OpVariable %_ptr_Function_v2float Function
               OpStore %param %int_0
         %37 = OpFunctionCall %v2float %rect_selection_i1_ %param
               OpStore %param_0 %int_1
         %39 = OpFunctionCall %v2float %rect_selection_i1_ %param_0
         %45 = OpAccessChain %_ptr_Input_float %vertPosition %uint_0
         %46 = OpLoad %float %45
         %47 = OpCompositeConstruct %v2float %46 %46
         %48 = OpExtInst %v2float %1 FMix %37 %39 %47
               OpStore %param_1 %int_3
         %51 = OpFunctionCall %v2float %rect_selection_i1_ %param_1
               OpStore %param_2 %int_2
         %54 = OpFunctionCall %v2float %rect_selection_i1_ %param_2
         %55 = OpAccessChain %_ptr_Input_float %vertPosition %uint_0
         %56 = OpLoad %float %55
         %57 = OpCompositeConstruct %v2float %56 %56
         %58 = OpExtInst %v2float %1 FMix %51 %54 %57
         %60 = OpAccessChain %_ptr_Input_float %vertPosition %uint_1
         %61 = OpLoad %float %60
         %62 = OpCompositeConstruct %v2float %61 %61
         %63 = OpExtInst %v2float %1 FMix %48 %58 %62
               OpStore %innerUv %63
               OpStore %param_3 %int_0
         %66 = OpFunctionCall %v2float %rect_position_i1_ %param_3
               OpStore %param_4 %int_1
         %68 = OpFunctionCall %v2float %rect_position_i1_ %param_4
         %69 = OpAccessChain %_ptr_Input_float %vertPosition %uint_0
         %70 = OpLoad %float %69
         %71 = OpCompositeConstruct %v2float %70 %70
         %72 = OpExtInst %v2float %1 FMix %66 %68 %71
               OpStore %param_5 %int_3
         %74 = OpFunctionCall %v2float %rect_position_i1_ %param_5
               OpStore %param_6 %int_2
         %76 = OpFunctionCall %v2float %rect_position_i1_ %param_6
         %77 = OpAccessChain %_ptr_Input_float %vertPosition %uint_0
         %78 = OpLoad %float %77
         %79 = OpCompositeConstruct %v2float %78 %78
         %80 = OpExtInst %v2float %1 FMix %74 %76 %79
         %81 = OpAccessChain %_ptr_Input_float %vertPosition %uint_1
         %82 = OpLoad %float %81
         %83 = OpCompositeConstruct %v2float %82 %82
         %84 = OpExtInst %v2float %1 FMix %72 %80 %83
               OpStore %innerPos %84
         %88 = OpAccessChain %_ptr_Function_float %innerPos %uint_0
         %89 = OpLoad %float %88
         %91 = OpAccessChain %_ptr_Function_float %innerPos %uint_1
         %92 = OpLoad %float %91
         %93 = OpFSub %float %float_1 %92
         %94 = OpCompositeConstruct %v2float %89 %93
         %95 = OpVectorTimesScalar %v2float %94 %float_2
         %96 = OpCompositeConstruct %v2float %float_1 %float_1
         %97 = OpFSub %v2float %95 %96
               OpStore %glslPos %97
        %103 = OpLoad %v2float %glslPos
        %105 = OpCompositeExtract %float %103 0
        %106 = OpCompositeExtract %float %103 1
        %107 = OpCompositeConstruct %v4float %105 %106 %float_0 %float_1
        %109 = OpAccessChain %_ptr_Output_v4float %_ %int_0
               OpStore %109 %107
        %112 = OpLoad %v2float %innerUv
               OpStore %uv %112
               OpReturn
               OpFunctionEnd
%rect_selection_i1_ = OpFunction %v2float None %10
          %i = OpFunctionParameter %_ptr_Function_int
         %13 = OpLabel
         %22 = OpLoad %int %i
         %24 = OpAccessChain %_ptr_Uniform_v2float %paint_coordinates %int_0 %22
         %25 = OpLoad %v2float %24
               OpReturnValue %25
               OpFunctionEnd
%rect_position_i1_ = OpFunction %v2float None %10
        %i_0 = OpFunctionParameter %_ptr_Function_int
         %16 = OpLabel
         %29 = OpLoad %int %i_0
         %30 = OpAccessChain %_ptr_Uniform_v2float %paint_coordinates %int_1 %29
         %31 = OpLoad %v2float %30
               OpReturnValue %31
               OpFunctionEnd
```

</details>

**Comment by HeroicKatora** [comment]
The original source code input to `shaderc` (not minimal but rather short):

<details>

```glsl
#version 450
#extension GL_EXT_scalar_block_layout : require
layout (location = 0) in vec2 vertPosition;
layout (location = 0) out vec2 uv;

layout (set = 0, binding = 0, std430) uniform PaintCoordinates {
  /* Positions in clockwise rotation starting at top-left. */
  // HACK(naga-1400) if we use arrays vec2[4], we violate Vulkan requirements of alignment.
  // if we use mat4x2 then naga miscalculates the layout and we end up getting overlapping fields.
  // So we use a poor mans array .. .. .
  mat4x2 rect_selection;
  mat4x2 rect_position;
} paint_coordinates;

// HACK(naga-1400)
vec2 rect_selection(int i) {
  return paint_coordinates.rect_selection[i];
}
vec2 rect_position(int i) {
  return paint_coordinates.rect_position[i];
}

void main() {
  // Top-left based position (coordinates as in the image package).
  vec2 innerUv = mix(
    mix(rect_selection(0), rect_selection(1), vertPosition.x),
    mix(rect_selection(3), rect_selection(2), vertPosition.x),
    vertPosition.y);

  vec2 innerPos = mix(
    mix(rect_position(0), rect_position(1), vertPosition.x),
    mix(rect_position(3), rect_position(2), vertPosition.x),
    vertPosition.y);

  vec2 glslPos = 2.0*vec2(innerPos.x, 1.0 - innerPos.y) - 1.0;

  gl_Position = vec4(glslPos, 0.0, 1.0);
  uv = innerUv;
}
```

**Comment by kvark** [comment]
Another layout issue, similar to #1414

**Comment by teoxoy** [comment]
This has been fixed by #1781. 

We have checks in place to check offsets don't overlap with other members but this was an issue with the `MatrixStride` decoration.

We were emitting a `MatrixStride` decoration with a value of 16 instead of 8 (making the size 64 instead of 32).
